text/0034-precomputed-entity-attributes.md

```diff
@@ -0,0 +1,186 @@
+functions for attribute values of type `ipaddr` or `decimal`).
+Today, `Entity::new()` never returns an error, and the `Entities` functions can
+return `EntitiesError` but not an evaluation error.
+- Conversely, `Entity::attr()` need not return an error anymore, as all of the
+attributes will already be evaluated. (If we are very concerned about backward
+compatibility, we could keep the current signature and just never return `Err`.
+Currently, this RFC is proposing we change the signature to not contain
+`Result`, since we're already making the related breaking change to
+`Entity::new()`.)
+
+Accepting these breaking changes allows us to give users the best-performing
+behavior by default.
+For alternatives that are less breaking, see [Alternatives](#alternatives).
+
+### Component 2: Construct entities using precomputed `Value`s
+
+Second, we add new constructors for `Entities` and `Entity` which take `Value`s
+instead of `RestrictedExpression`s.
+These constructors would work without throwing evaluation errors, in contrast
+to the existing constructors after the changes described above.
+This addresses the third point in [Motivation](#motivation).
+
+This change requires that we expose `Value` in our public API in some manner
+(probably via a wrapper around the Core implementation, as we do for many other
+Core types).
+Note that today, `EvalResult` plays the role of `Value` in our public API, but
+conversion between `EvalResult` and `Value` is relatively expensive (`O(n)`).
+We could either:
+
+- expose `Value` but leave `EvalResult` alone. This provides maximum backwards
+compatibility, but might cause confusion, as `Value` and `EvalResult`
+would be conceptually similar, but slightly different and not interchangeable.
+- consolidate `EvalResult` and `Value` into a single representation, called
+`Value` and based on the Core `Value`. This provides maximum API cleanliness
+but requires the most breaking changes.
+- expose `Value`, but keep `EvalResult` as a type synonym for `Value`. This
+is almost the same as the above option, but keeps the `EvalResult` type name
+around, even though it makes breaking changes to the structure of
+`EvalResult`. This might reduce confusion for users who are migrating, as
+they only have to adapt to the changes to `EvalResult` and not rename
+`EvalResult` to `Value` in their own code; but, it might increase confusion
+for future users, who will see two names for the same thing.
+- not expose `Value` and just use `EvalResult`, even in the new constructors.
+This provides maximum backwards compatibility, like the first option, and
+also avoids the possible confusion generated by the first option, but comes
+at a notable performance cost (due to the inefficient
+`EvalResult`-to-`Value` conversion required under the hood) and might
+generate a different kind of confusion as `EvalResult` is an awkward name in
+the context of these constructors.
+
+This RFC currently proposes we take the third option.
+
```

Comment:
Same question as above: What happens when you try to compile existing code, which will use `EvalResult` to use the new API? I think you will get a type error that is easily fixed?

Reply:
See my answer above.  You will get a compile error and have to adapt to the new structure, depending on how you are using the `EvalResult`.  I don't know if anyone is actually using `EvalResult`, and if so, how they are using it.  If they are just printing the result to a string/console (as the CLI does), then the fix is minimal or zero, as the new version will be equally formattable with the Rust `Display` trait.

Reply:
Resolving this thread now that this component is no longer in the main proposal.